File: benches/benchmark.rs
//! Benchmark to test the speed of opening files

use std::fs::File;
use std::io::prelude::*;
use std::io::BufWriter;
use std::time::{Duration, Instant};

use pdbtbx::*;

fn main() {
    // Setup the data needed
    let pdb_names = vec![
        ("small", "example-pdbs/1ubq.pdb"),
        ("medium", "example-pdbs/1yyf.pdb"),
        ("big", "example-pdbs/pTLS-6484.pdb"),
        ("huge", "example-pdbs/large.pdb"),
        // ("extreme", "example-pdbs/1htq.pdb"),
    ];
    let mmcif_names = vec![
        ("small", "example-pdbs/1ubq.cif"),
        ("medium", "example-pdbs/1yyf.cif"),
        ("big", "example-pdbs/pTLS-6484.cif"),
    ];
    let mut models = Vec::with_capacity(pdb_names.len());
    let parser = ReadOptions::default()
        .set_level(StrictnessLevel::Loose)
        .set_format(Format::Pdb)
        .clone();
    for (name, path) in &pdb_names {
        models.push((*name, parser.read(path).unwrap().0));
    }
    let mut results = Vec::new();

    // Do the benchmarking
    results.extend(measure_multiple(bench_open, &pdb_names, "Open PDB"));
    results.extend(measure_multiple(bench_open, &mmcif_names, "Open mmCIF"));
    results.extend(measure_multiple(
        bench_transformation,
        &models,
        "Transformation",
    ));
    results.extend(measure_multiple(bench_remove, &models, "Remove"));
    results.extend(measure_multiple(bench_iteration, &models, "Iteration"));
    results.extend(measure_multiple(bench_validate, &models, "Validate"));
    results.extend(measure_multiple(bench_renumber, &models, "Renumber"));
    results.extend(measure_multiple(bench_clone, &models, "Clone"));
    results.extend(measure_multiple(bench_save_pdb, &models, "Save PDB"));
    results.extend(measure_multiple(bench_save_mmcif, &models, "Save mmCIF"));

    // Save the results to a csv
    let file = File::create("dump/benchmark_results.csv").unwrap();
    let mut sink = BufWriter::new(file);
    sink.write_all("Name,Average(ns),StandardDeviation(ns),Runs\n".as_bytes())
        .unwrap();
    for item in results {
        sink.write_fmt(format_args!(
            "{},{},{},{}\n",
            item.0, item.1, item.2, item.3
        ))
        .unwrap();
    }
    sink.flush().unwrap();
}

fn bench_open(filename: &str) {
    let (_pdb, _errors) = ReadOptions::default()
        .set_level(StrictnessLevel::Loose)
        .read(filename)
        .unwrap();
}

fn bench_transformation(mut pdb: PDB) {
    let transformation = TransformationMatrix::rotation_x(90.0);
    pdb.apply_transformation(&transformation);
}

fn bench_remove(mut pdb: PDB) {
    pdb.remove_atoms_by(|atom| atom.serial_number() % 2 == 0);
}

fn bench_iteration(pdb: PDB) {
    let mut _average = 0.0;
    for atom in pdb.atoms() {
        _average += atom.b_factor();
    }
    _average /= pdb.atom_count() as f64;
}

fn bench_validate(pdb: PDB) {
    let _ = validate(&pdb);
}

fn bench_renumber(mut pdb: PDB) {
    pdb.renumber();
}

fn bench_clone(pdb: PDB) {
    let _copy = pdb;
}

fn bench_save_pdb(pdb: PDB) {
    save(&pdb, "dump/dump.pdb", StrictnessLevel::Loose).unwrap();
}

fn bench_save_mmcif(pdb: PDB) {
    save(&pdb, "dump/dump.cif", StrictnessLevel::Loose).unwrap();
}

fn measure_multiple<T: Clone>(
    function: fn(T),
    subjects: &[(&str, T)],
    description: &str,
) -> Vec<(String, u128, u128, usize)> {
    let mut output = Vec::with_capacity(subjects.len());
    for (name, item) in subjects {
        output.push(measure(function, item, &format!("{description} - {name}")));
    }
    output
}

fn measure<T: Clone>(
    function: fn(T),
    subject: &T,
    description: &str,
) -> (String, u128, u128, usize) {
    let mut times = Vec::new();
    function(subject.clone());
    let start = Instant::now();
    let mut now;

    for _ in 0..5 {
        let clone = subject.clone();
        now = Instant::now();
        function(clone);
        times.push(now.elapsed());
    }

    let average = start.elapsed().checked_div(5).unwrap();

    // Lets run for 3 more seconds, including cloning of the subject
    for _ in 0..3_000_000_000 / average.as_nanos() {
        let clone = subject.clone();
        now = Instant::now();
        function(clone);
        times.push(now.elapsed());
    }

    let average = times
        .iter()
        .fold(Duration::new(0, 0), |total, item| {
            total.checked_add(*item).unwrap()
        })
        .checked_div(times.len() as u32)
        .unwrap();

    let mut deviation = 0.0;
    for run in &times {
        let difference = run.as_nanos() as f64 - average.as_nanos() as f64;
        deviation += difference * difference;
    }
    deviation /= times.len() as f64;
    deviation = deviation.sqrt();

    (
        description.to_string(),
        average.as_nanos(),
        deviation as u128,
        times.len(),
    )
}

File: benches/benchmark.py
from iotbx.pdb import hierarchy as Hierarchy
import timeit
import time
import math
from transformation import matrix_rotate_3d_x, matrix_3d_to_4x4
import numpy as np
import copy


def apply_3d_matrix(position, matrix):
    """
    Applies a 4x4 affine transformation matrix to the given 3D position
    """
    (x, y, z) = position
    new_position = np.dot(matrix,
                          np.matrix([[x], [y], [z], [1]]))
    return (new_position.item(0), new_position.item(1), new_position.item(2))


def transform_protein(hierarchy, transformation, a=None, b=None, c=None):
    """
    Transform the given hierarchy by the given affine transformation matrix
    Returns: nothing, it edits the hierarchy itself
    """
    mod = False
    if a != None:
        mod = True
        assert(b != None)
        assert(c != None)

    for atom in hierarchy.atoms():
        new_position = apply_3d_matrix(atom.xyz, transformation)
        if mod:
            new_position = (
                new_position[0] % a, new_position[1] % b, new_position[2] % c)
        atom.set_xyz(new_position)


def open_pdb(filename):
    Hierarchy.input(file_name=filename)


def transformation(pdb):
    transform_protein(pdb.hierarchy, matrix_3d_to_4x4(
        matrix_rotate_3d_x(90)))


def remove(pdb):
    for atom in pdb.hierarchy.atoms():
        if float(atom.serial) % 2 == 0:
            atom.set_element("DL")

    result = pdb.hierarchy.select(
        pdb.hierarchy.atom_selection_cache().selection("not (element DL)"))


def iteration(pdb):
    average = 0
    for atom in pdb.hierarchy.atoms():
        average += atom.b
    average = average / pdb.hierarchy.atoms_size()


def iteration_build_in(pdb):
    atoms = pdb.hierarchy.atoms()
    average = atoms.extract_b().min_max_mean().mean


def renumber(pdb):
    for chain in pdb.hierarchy.chains():
        chain.id = chain.id[0]

    serial_number = 1
    for atom in pdb.hierarchy.atoms():
        atom.set_serial(serial_number)
        serial_number += 1

    serial_number = 1
    for rg in pdb.hierarchy.residue_groups():
        rg.resseq = serial_number
        serial_number += 1


def clone(pdb):
    copy.deepcopy(pdb)

# No validate


def save(pdb):
    f = open("dump.pdb", "w")
    f.write(pdb.hierarchy.as_pdb_string())
    f.close()


def time_function(fn, subject, name):
    times = []
    start = time.time()
    for i in range(0, 5):
        subject_copy = copy.deepcopy(subject)
        start_inner = time.time()
        result = fn(subject_copy)
        end_inner = time.time()
        times.append(end_inner - start_inner)
    end = time.time()

    runs = math.ceil(5 / (end - start))
    for i in range(0, runs):
        subject_copy = copy.deepcopy(subject)
        start_inner = time.time()
        result = fn(subject_copy)
        end_inner = time.time()
        times.append(end_inner - start_inner)

    avg = sum(times) / (5 + runs)
    stdev = np.std(times)
    print(name + "\t" + str(avg) + "\t" + str(stdev) + "\t" + str(5 + runs))


def time_function_multiple(fn, subjects, global_name):
    for (name, subject) in subjects:
        time_function(fn, subject, global_name + "\t" + name)


names = [
    ("small", "example-pdbs/1ubq.pdb"),
    ("medium", "example-pdbs/1yyf.pdb"),
    ("big", "example-pdbs/pTLS-6484.pdb"),
]

proteins = [
    ("small", Hierarchy.input(file_name="example-pdbs/1ubq.pdb")),
    ("medium", Hierarchy.input(file_name="example-pdbs/1yyf.pdb")),
    ("big", Hierarchy.input(file_name="example-pdbs/pTLS-6484.pdb")),
]

time_function_multiple(open_pdb, names, "open")
time_function_multiple(transformation, proteins, "transformation")
time_function_multiple(remove, proteins, "remove")
time_function_multiple(iteration, proteins, "iteration")
time_function_multiple(iteration_build_in, proteins, "iteration_build_in")
time_function_multiple(renumber, proteins, "renumber")
time_function_multiple(clone, proteins, "clone")
time_function_multiple(save, proteins, "save")
